googledataprocauthenticator/google.py

```diff
@@ -0,0 +1,319 @@
+            "authenticate. Run `gcloud auth application-default login` acquire new user "\
+            "credentials to use for Application Default Credentials."
+        )
+        raise new_exc
+
+def get_credentials_for_account(account, scopes_list):
+    """Load all of user's credentialed accounts with ``gcloud auth describe ACCOUNT`` command.
+
+    Args:
+        account (str): user credentialed account to return credentials for
+        scopes_list (Sequence[str]): list of scopes to include in the credentials.
+    
+    Returns:
+        google.oauth2.credentials.Credentials: The constructed credentials
+
+    Raises:
+        ValueError: If `gcloud auth describe ACCOUNT --format json` returns json not in the expected format.
+        sparkmagic.livyclientlib.BadUserConfigurationException: if account is not set or user needs to run 
+        gcloud auth login or if gcloud is not installed. 
+    """
+    if os.name == "nt":
+        command = _CLOUD_SDK_WINDOWS_COMMAND
+    else:
+        command = _CLOUD_SDK_POSIX_COMMAND
+    try:
+        describe_account_command = ("auth", "describe", account, '--format', 'json')
+        command = (command,) + describe_account_command
+        account_json = subprocess.check_output(command, stderr=subprocess.STDOUT)
+        account_describe = load_json_input(account_json)
+        return Credentials.from_authorized_user_info(account_describe, scopes=scopes_list)
+    except ValueError: 
+        raise
+    except (OSError) as caught_exc:
+        new_exc = BadUserConfigurationException(
+            "Gcloud is not installed. Install the Google Cloud SDK." 
+        )
+        raise new_exc
+    except (subprocess.CalledProcessError, IOError) as caught_exc:
+        new_exc = BadUserConfigurationException(
+            "Failed to obtain access token. Run `gcloud auth login` in your command line "\
+            "to authorize gcloud to access the Cloud Platform with Google user credentials "\
+            "to authenticate. Run `gcloud auth application-default login` to acquire new "\
+            "user credentials to use for Application Default Credentials.")
+        raise new_exc
+
+
+def get_cluster_pool(project_id, region, client):
+    #filter format: status.state = ACTIVE AND clusterName = mycluster AND labels.env = staging AND labels.starred = \*
+    cluster_pool = set()
+    for cluster in client.list_clusters(project_id, region, 'status.state = ACTIVE'):
+        #check component gateway is enabled
+        if (len(cluster.config.endpoint_config.http_ports.values()) != 0):
+            action_list = list()
+            for action in cluster.config.initialization_actions:
+                #check if init action is the livy action with a region that matches the regex pattern [a-z0-9-]+
+                is_livy_action = re.search("gs://goog-dataproc-initialization-actions-[a-z0-9-]+/livy/livy.sh", bucket) is not None                    
```

Comment:
This regex means that users who fork the init action can't use this feature.

Instead, could we just assume that Livy is enabled, and add clusters to the set anyway. We can then try to make requests to `https://cg-host/livy` later on and catch exceptions if Livy isn't installed

Reply:
I'm conflicted about this. We don't officially support forked init actions with Component Gateway right now: for Rstudio and Hue, we'll only expose the URL if they're using the canned init action. On the other hand, that is more of a soft block since they can still always try the URL to see if the component is there.

It looks like the file was changed so I can't see the original context of this, but if they're manually specifying the cluster name it's fine to try first and then fail since presumably they believe that cluster should have Livy. When we actually show a list of available clusters (or when we're selecting by labels) I'd be more comfortable restricting it to clusters we know are using the canned init action.

Reply:
I moved the regex stuff to a different filter by label branch so this one could just be setup but the full function this comment here is on is the get_cluster_pool() here: https://github.com/GoogleCloudDataproc/dataprocmagic/blob/filter-by-label/googledataprocauthenticator/google.py#L148

Reply:
Ah okay, got it: re component gateway